### DISCUSSION TITLE
Need to be able to wake up Tasmota from DeepSleep by using a touch pin or button press.

### DISCUSSION THREAD

**Discussion by LucidEye** [arendst/Tasmota#17972] [Ideas]
Hello,

First, I would like to thank all of the devs and contributors who have made Tasmota such an incredibly useful and feature-rich project... I've only been tinkering with ESP32s and Tasmota for 2 months, and I'm constantly amazed at how much they have managed to cram into this firmware.

My main goal with using Tasmota at the moment is to build several agricultural data loggers & monitoring sensors that will be used to automate an irrigation system for a large outdoor vegetable garden.  I will be using solar panels and batteries to power the ESP32 units, which means I will be using the DeepSleep feature.  I have already implemented the DeepSleep "loop stop" switch as described here... https://tasmota.github.io/docs/DeepSleep/

What I really need is a way to force a wake-up from DeepSleep mode by using an external event, such as a push button, or more preferably one of the ESP32 capacitive touch pins.  I am unable to find any documentation or user projects that do this.  I have been told in the Discord discussion that Tasmota has no support for external wake-up events?????  Is this true????

For in-the-field troubleshooting and servicing it is impractical to have to wait for the DeepSleep cycle to finish on it's own before the units can be accessed via the web interface.  It is also impractical to simply power-cycle the units since they will be completely sealed in waterproof cases.  I will have a switch accessible on the outside of the case to disable the DeepSleep loop, but I also need a second external switch or button or touch pin to force a wake-up so that the unit is then prevented from entering DeepSleep by the loop-stop switch rather than having to wait for the DeepSleep cycle to finish on it's own.

How can this be done?  Is it even possible with the current release?  If not, can this be an Enhancement Request?

Thank you for any helpful info.

**Comment by pkkrusty** [comment]
Made any progress with this?

**Reply by LucidEye** [reply]
No luck so far :-(

**Reply by pkkrusty** [reply]
As far as I can tell, GPIO16 pulls Reset to ground to initiate a wake up from deepsleep. So doing the same thing manually with a button should wake it up, right? I haven't tried it, but may play with it this weekend. ![image](https://user-images.githubusercontent.com/79770016/221123759-f9e797da-df11-46c9-b9cc-0fe02e82a71d.png)

See this stack exchange thread: https://arduino.stackexchange.com/questions/71207/esp8266-deep-sleep-wake-up

**Reply by barbudor** [reply]
That's for esp8266 but the topic here is esp32

Beside that, Deepsleep in Tasmota is intended to be used for periodic wake up to send sensors to the backend, not to be woke up on demand

Let's say that you configure such button to drag RST low (most dev boards already have one)
You enter Deepsleeptime with along value of 1 year because you don't want the esp to wake on it's own

Once the esp is aslepp you press the button
The esp check the time and say 'OMG why did I woke so early, I was supposed to wake up only in a year. Let immediately go to sleep before doing anything else'
And your screwed

I know that on ESP32 some work has been done to allow usage of the ULP inside the ESP32 so there's may be things to look at here (search PR and discussions)

**Reply by LucidEye** [reply]
Thanks for explaining that a bit, but in my use-case it would not "go right back to sleep" because I would be using the forced wake up in conjunction with the "stop sleep" loop switch as I described in my original post.  The reason I wanted to use a touch pin was because I wanted to attach the touch pin to the metal housing of the switch I was going to use as the stop-sleep loop switch, thus creating 2 switches in one.... touching the stop loop switch would bring the esp32 out of sleep mode, and actually flipping the switch would prevent it from returning to sleep mode once woken up.... at least that was my idea.  But now it looks like the workaround is to simply place a momentary external switch on the power source so that I can just reboot the esp32 after switching on the stop-sleep loop switch.  Rebooting will bring it out of sleep not an elegant solution, but it should work more or less the same way as my original idea.

**Reply by barbudor** [reply]
I fear that the "stop sleep" GPIO would be tested only in case of a full wakeup of Tasmota at the time it would go back for a next DeepSleep period
In the above description I'm taking about a test done early in the boot sequence which may ignore the "stop sleep" GPIO.

Please experiment and let us know
The DeepSleep feature was originally designed for ESP8266 and there is much more possibilities with the ESP32
There is room for improvement.
Also have a look to ULP as Jason added the link to it

**Comment by Jason2866** [comment]
There is an example to do with Berry and ULP here
https://github.com/Staars/berry-examples/blob/main/ulp_examples/ulp_gpio_wake.py

**Reply by LucidEye** [reply]
Thank you for the example, but this seems WAY over-complicated for just waking up from sleep with one touch pin.  I am not a programmer, and I know nothing about berry script, and I simply do not have the time to learn a whole new scripting language before I want to get these working and installed in my garden in a few weeks.  For now I am just going to add a second switch that will power-cycle the esp32 to force it out of sleep mode.
